### PR TITLE
chore: create Field group config and refactor ActionCreator

### LINKED ISSUE #16938: ActionCreator/index.tsx refactoring
There may be some scope to improve some of the logic this file contains, even after the config object with all the functions, fields, and configurations is created.

This is a placeholder task for any sort of housekeeping this file needs. As per my analysis, 2 places can be broken down into simpler code
1. getIntegrationOptionsWithChildren function
2. useIntegrationsOptionTree function
3. action creator component
4. move utils/constants to their own file (Covered by issue https://app.zenhub.com/workspace/o/appsmithorg/appsmith/issues/16932)

This will be removed if there is no more scope for refactoring

### LINKED ISSUE #17347: Create config object for FieldGroup

### PR DESCRIPTION
**Pull request by rimildeyjsr**
## Description

As part of the action selector refactor, we are making the following changes -
- Change Fields -> FieldGroup
- Create Field group config
- Add new types
- Refactor action creator component

Fixes #16938 
Fixes #17347

Untitled-2022-09-27-1046

## Type of change

- Refactor

## How Has This Been Tested?
- Manually
- Jest test cases

### Test plan
- [ ] https://github.com/appsmithorg/TestSmith/issues/2099
- [ ] https://github.com/appsmithorg/TestSmith/issues/2100
- [ ] https://github.com/appsmithorg/TestSmith/issues/2101
- [ ] https://github.com/appsmithorg/TestSmith/issues/2102
- [ ] https://github.com/appsmithorg/TestSmith/issues/2103
- [ ] https://github.com/appsmithorg/TestSmith/issues/2104
- [ ] clearStore function should behave as is and should not default to a value
- [ ] https://github.com/appsmithorg/TestSmith/issues/2105
- [ ] https://github.com/appsmithorg/TestSmith/issues/2106
- [ ] https://github.com/appsmithorg/TestSmith/issues/2107
- [ ] https://github.com/appsmithorg/TestSmith/issues/2108
- [ ] https://github.com/appsmithorg/TestSmith/issues/2109
- [ ] https://github.com/appsmithorg/TestSmith/issues/2110
- [ ] watchPosition function should behave as is and should not default to a value
- [ ] clearWatch function should behave as is and should not default to a value
- [ ] https://github.com/appsmithorg/TestSmith/issues/2111

### Dev activity:
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### QA activity:
- [ ] Test plan has been approved by relevant developers
- [ ] Test plan has been peer reviewed by QA
- [ ] Cypress test cases have been added and approved by either SDET or manual QA
- [ ] Organized project review call with relevant stakeholders after Round 1/2 of QA
- [ ] Added `Test Plan Approved` label after reveiwing all Cypress test

### CODE REVIEW

**Review by ohansFavour — COMMENTED** [review]
@rimildeyjsr As we make incremental changes to the Action selector structure, we should also make these changes to the Notion doc detailing steps to Add/Remove actions in Appsmith.

**Comment by rimildeyjsr** [comment]
> @rimildeyjsr As we make incremental changes to the Action selector structure, we should also make these changes to the Notion doc detailing steps to Add/Remove actions in Appsmith.

@ohansFavour - I have docs in my pipeline - once the design changes are done, I will update them

File: app/client/src/components/editorComponents/ActionCreator/Field/FieldConfig.ts [hunk 1]
```diff
@@ -5,26 +5,26 @@ import {
   NAVIGATION_TARGET_FIELD_OPTIONS,
   RESET_CHILDREN_OPTIONS,
   ViewTypes,
-} from "./constants";
-import { ALERT_STYLE_OPTIONS } from "../../../ce/constants/messages";
-import { ActionType, AppsmithFunctionConfigType, FieldProps } from "./types";
+} from "../constants";
+import { ALERT_STYLE_OPTIONS } from "ce/constants/messages";
```
Comment (ohansFavour) [inline app/client/src/components/editorComponents/ActionCreator/Field/FieldConfig.ts:9]: Please use @appsmith/*

File: app/client/src/components/editorComponents/ActionCreator/Field/FieldConfig.ts [hunk 2]
```diff
@@ -5,26 +5,26 @@ import {
...
+import store from "store";
+import { getPageList } from "selectors/entitiesSelector";
+import { ACTION_TRIGGER_REGEX } from "../regex";
 import { TreeDropdownOption } from "design-system";
 
 export const FIELD_CONFIG: AppsmithFunctionConfigType = {
   [FieldType.ACTION_SELECTOR_FIELD]: {
     label: (props: FieldProps) => props.label || "",
-    options: (props: FieldProps) => props.integrationOptionTree,
+    options: (props: FieldProps) => props.queriesAndJsObjectsOption,
```
Comment (ohansFavour) [inline app/client/src/components/editorComponents/ActionCreator/Field/FieldConfig.ts:27]: Do Actions fall under these options too?.

File: app/client/src/components/editorComponents/ActionCreator/Field/FieldConfig.ts [hunk 3]
```diff
@@ -68,7 +68,7 @@ export const FIELD_CONFIG: AppsmithFunctionConfigType = {
         case AppsmithFunction.resetWidget:
           defaultParams = `"",true`;
           break;
-        case AppsmithFunction.postMessage:
+        case AppsmithFunction.postWindowMessage:
```
Comment (ohansFavour) [inline app/client/src/components/editorComponents/ActionCreator/Field/FieldConfig.ts:71]: Why `postWindowMessage`?, Are we accounting for posting to Iframes too?

File: app/client/src/components/editorComponents/ActionCreator/index.tsx [hunk 4]
```diff
@@ -165,250 +74,208 @@ function getFieldFromValue(
...
+      path && entity.meta && entity.meta[path] && entity.meta[path].arguments;
     fields.push({
-      field: FieldType.CALLBACK_FUNCTION_FIELD,
+      field: FieldType.ACTION_SELECTOR_FIELD,
+      getParentValue,
+      value,
+      args: argsProps ? argsProps : [],
     });
+    if (argsProps && argsProps.length > 0) {
+      for (let i = 0; i < argsProps.length; i++) {
```
Comment (ohansFavour) [inline app/client/src/components/editorComponents/ActionCreator/index.tsx:222]: Use "for let argsProp of argsProps"

File: app/client/src/components/editorComponents/ActionCreator/utils.ts [hunk 5]
```diff
@@ -143,3 +144,16 @@ export const enumTypeGetter = (
     return arg ? arg.trim() : defaultValue;
   }
 };
+
+export const isValueValidURL = (value: string) => {
```
Comment (ohansFavour) [inline app/client/src/components/editorComponents/ActionCreator/utils.ts:148]: Why do we need isValueValidURL and isValidURL, when they both accept a string? Can't we empower isValidURL to do this additional check?. Can you also add a comment here, it is difficult to understand the reason for the additional checks in isValueValidURL.

File: app/client/src/components/editorComponents/ActionCreator/utils.ts [hunk 6]
```diff
@@ -7,6 +7,7 @@ import {
   getDynamicBindings,
   isDynamicValue,
 } from "../../../utils/DynamicBindingUtils";
+import { isValidURL } from "../../../utils/URLUtils";
```
Comment (ohansFavour) [inline app/client/src/components/editorComponents/ActionCreator/utils.ts:10]: Use utils/*

File: app/client/src/components/editorComponents/ActionCreator/types.ts [hunk 7]
```diff
@@ -73,15 +73,15 @@ export type FieldProps = {
   widgetOptionTree: TreeDropdownOption[];
   modalDropdownList: TreeDropdownOption[];
   pageDropdownOptions: TreeDropdownOption[];
-  integrationOptionTree: TreeDropdownOption[];
+  queriesAndJsObjectsOption: TreeDropdownOption[];
```
Comment (ohansFavour) [inline app/client/src/components/editorComponents/ActionCreator/types.ts:76]: Once again if these options include Api/actions, let's use a better name here. This appears to me as strictly for DB queries and JS functions

File: app/client/src/components/editorComponents/ActionCreator/index.tsx [hunk 8]
```diff
@@ -441,75 +308,118 @@ function useModalDropdownList() {
...
           api.config,
           plugins[(api as any).config.datasource.pluginId],
           api.config.pluginType === PluginType.API,
         ),
       } as TreeDropdownOption);
     });
+
     queries.forEach((query) => {
-      (option.children as TreeDropdownOption[]).push({
+      (queriesOption.children as TreeDropdownOption[]).push({
```
Comment (ohansFavour) [inline app/client/src/components/editorComponents/ActionCreator/index.tsx:383]: queryOptions?

File: app/client/src/components/editorComponents/ActionCreator/types.ts (hunk 7 above)
Comment (rimildeyjsr) [reply app/client/src/components/editorComponents/ActionCreator/types.ts:76]: renamed

File: app/client/src/components/editorComponents/ActionCreator/Field/FieldConfig.ts (hunk 3 above)
Comment (ohansFavour) [reply app/client/src/components/editorComponents/ActionCreator/Field/FieldConfig.ts:71]: If so, don't you think that "Window" in the name of the function will be confusing?

File: app/client/src/components/editorComponents/ActionCreator/Field/FieldConfig.ts [hunk 9]
```diff
@@ -329,15 +329,15 @@ export const FIELD_CONFIG: AppsmithFunctionConfigType = {
   },
   [FieldType.ON_SUCCESS_FIELD]: {
     label: () => "",
-    options: (props: FieldProps) => props.integrationOptionTree,
+    options: (props: FieldProps) => props.apisQueriesAndJsObjectsOption,
```
Comment (ohansFavour) [inline app/client/src/components/editorComponents/ActionCreator/Field/FieldConfig.ts:332]: In my opinion, "integrationOption" is a better name here

File: app/client/src/components/editorComponents/ActionCreator/Field/FieldConfig.ts (hunk 1 above)
Comment (rimildeyjsr) [reply app/client/src/components/editorComponents/ActionCreator/Field/FieldConfig.ts:9]: done

[the remark this answers is not in the document]
Comment (rimildeyjsr) [reply app/client/src/components/editorComponents/ActionCreator/Field/FieldConfig.ts:9]: done

File: app/client/src/components/editorComponents/ActionCreator/Field/FieldConfig.ts (hunk 3 above)
Comment (rimildeyjsr) [reply app/client/src/components/editorComponents/ActionCreator/Field/FieldConfig.ts:71]: @ohansFavour - we went back and forth a lot on the name and rationalized that iframe is also a window so it's fine. I will look into a different name if needed in the V2 version of this

File: app/client/src/components/editorComponents/ActionCreator/Field/FieldConfig.ts (hunk 9 above)
Comment (rimildeyjsr) [reply app/client/src/components/editorComponents/ActionCreator/Field/FieldConfig.ts:332]: done

**Review by ohansFavour — APPROVED** [review]

**Comment by rimildeyjsr** [comment]
This feature is moved to https://github.com/appsmithorg/appsmith/pull/21582

**Pull request closed without merging**